src/main/java/org/zalando/jackson/datatype/money/MonetaryAmountDeserializer.java

```diff
@@ -1,24 +1,54 @@
 package org.zalando.jackson.datatype.money;
 
+import com.fasterxml.jackson.core.JsonParseException;
 import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonToken;
 import com.fasterxml.jackson.databind.DeserializationContext;
 import com.fasterxml.jackson.databind.JsonDeserializer;
 
+import javax.money.CurrencyUnit;
 import javax.money.MonetaryAmount;
 import java.io.IOException;
+import java.math.BigDecimal;
+
+import static java.lang.String.format;
 
 public final class MonetaryAmountDeserializer<M extends MonetaryAmount> extends JsonDeserializer<M> {
 
     private final MonetaryAmountFactory<M> factory;
+    private final FieldNames names;
 
-    public MonetaryAmountDeserializer(final MonetaryAmountFactory<M> factory) {
+    public MonetaryAmountDeserializer(final MonetaryAmountFactory<M> factory, final FieldNames names) {
         this.factory = factory;
+        this.names = names;
     }
 
     @Override
     public M deserialize(final JsonParser parser, final DeserializationContext context) throws IOException {
-        final MoneyNode node = context.readValue(parser, MoneyNode.class);
-        return factory.create(node.getAmount(), node.getCurrency());
+        BigDecimal amount = null;
+        CurrencyUnit currency = null;
+
+        while (parser.nextToken() != JsonToken.END_OBJECT) {
+            final String field = parser.getCurrentName();
+
+            parser.nextToken();
+
+            if (field.equals(names.getAmount())) {
+                amount = parser.getDecimalValue();
+            } else if (field.equals(names.getCurrency())) {
+                currency = context.readValue(parser, CurrencyUnit.class);
+            }
```

Comment:
I don't think this will work for unknown fields, you would have to skip the (possibly complex) value for these fields using

    parser.skipChildren();